### PR TITLE
Pytorch/MXNet plugin - use dictionary of categories

### PR DESCRIPTION
**Pull request by klecki**
Split from #240.
Explicit naming of each output for DALIGenericIterator in Pytorch and MxNet.
MxNet requires additional grouping into `data` and `label` categories of input tensors, and label tensors are squeezed.

Iterator in Pytorch will return dictionary {"name" : tensor}, Iterator in MxNet will return DataBatch with names preserved in provide_data/provide_label as mx.io.DataDesc.

### CODE REVIEW

File: dali/python/nvidia/dali/plugin/pytorch.py [hunk 1]
```diff
@@ -155,6 +171,7 @@ def __next__(self):
         # Change index for double buffering
         self._current_data_batch = (self._current_data_batch + 1) % 2
         self._counter += self._num_gpus * self.batch_size
+        # TODO(klecki): convert to tuple for compability with "data" "labels"
```
Comment (JanuszL) [inline dali/python/nvidia/dali/plugin/pytorch.py]: Do we need to do that?

File: dali/python/nvidia/dali/plugin/pytorch.py [hunk 2]
```diff
@@ -107,45 +112,56 @@ def __next__(self):
...
+            category_shapes = dict()
+            for category, output_list in category_outputs.items():
+                category_tensors[category] = [x.as_tensor() for x in output_list]
+                category_shapes[category] = [x.shape() for x in category_tensors[category]]
+
             # Change label shape from [batch_size, 1] to [batch_size]
-            labels = [x.as_tensor() for x in out_labels]
-            for l in labels:
-                l.squeeze()
+            if "labels" in self._output_categories:
```
Comment (Kh4L) [inline dali/python/nvidia/dali/plugin/pytorch.py]: I think that we should remove this hacky "squeeze only if the output is named `labels`" and leave it to the user to squeeze the output or not, outsite of the plugin.

File: docs/examples/pytorch/main.py [hunk 3]
```diff
@@ -292,8 +292,8 @@ def train(train_loader, model, criterion, optimizer, epoch):
     end = time.time()
 
     for i, data in enumerate(train_loader):
-        input = data[0][0][0]
-        target = data[0][1][0].cuda().long()
+        input = data[0]["data"][0]
+        target = data[0]["label"][0].cuda().long()
```
Comment (Kh4L) [inline docs/examples/pytorch/main.py]: This would not work, because it should be called `labels` and not `label`.

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 4]
```diff
@@ -99,6 +97,12 @@ def __init__(self,
         self._data_batches = [[None] for i in range(self._num_gpus)]
         self._counter = 0
         self._current_data_batch = 0
+        self._output_names_map = [x[0] for x in output_map]
+        self._output_categories_map = [x[1] for x in output_map]
+        self._output_categories = {DATA_TAG, LABEL_TAG]}
```
Comment (Kh4L) [inline dali/python/nvidia/dali/plugin/mxnet.py]:
```suggestion
        self._output_categories = {[DATA_TAG, LABEL_TAG]}
```

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 5]
```diff
@@ -254,8 +277,9 @@ def __init__(self,
                  label_name='softmax_label',
                  data_layout='NCHW',
                  fill_last_batch=True):
-        super(DALIClassificationIterator, self).__init__(pipelines, ["data", "label"], size,
-                                                         data_name       = data_name,
-                                                         label_name      = label_name,
+        super(DALIClassificationIterator, self).__init__(pipelines,
+                                                         [data_name, DALIClassificationIterator.DATA_TAG,
+                                                          label_name, DALIClassificationIterator.LABEL_TAG],
```
Comment (Kh4L) [inline dali/python/nvidia/dali/plugin/mxnet.py]:
```suggestion
                                                          [(data_name, DALIClassificationIterator.DATA_TAG),
                                                           (label_name, DALIClassificationIterator.LABEL_TAG)],
```

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 6]
```diff
@@ -130,37 +141,56 @@ def __next__(self):
...
+                [(x.shape(), np.dtype(x.dtype())) for x in category_tensors[DALIGenericIterator.DATA_TAG]]
+            # For labels we squeeze the tensors
+            category_tensors[DALIGenericIterator.LABEL_TAG] = \
+                [x.as_tensor() for x in category_outputs[DALIGenericIterator.LABEL_TAG]]
+            for label in category_tensors[DALIGenericIterator.LABEL_TAG]:
+                label.squeeze()
+            category_info[DALIGenericIterator.LABEL_TAG] = \
+                [(x.shape(), np.dtype(x.dtype())) for x in category_tensors[DALIGenericIterator.LABEL_TAG]]
+
             # If we did not yet allocate memory for that batch, do it now
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Mhhh. But bbox encoder should guard us against different sizes at each iteration.

[the remark this answers is not in the document]
Comment (awolant) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Yes, encoder does, so for SSD it is not the problem. But in this COCO detection pipeline example, we return raw COCO boxes straight from Crop/Slice combo.

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 7]
```diff
@@ -60,16 +60,16 @@ class DALIGenericIterator(object):
...
-    output_map : list of str
-                 List of strings (either "data" or "label") which maps
-                 the output of DALI pipeline to proper field in MXNet's
-                 DataBatch
+    output_map : list of (str, str)
+                 List of pairs (output_name, tag) which maps consecutive
+                 outputs of DALI pipelines to proper field in MXNet's
+                 DataBatch.
+                 tag is one of DALIGenericIterator.DATA_TAG
+                 and DALIGenericIterator.LABEL_TAG mapping given output
```
Comment (JanuszL) [inline dali/python/nvidia/dali/plugin/mxnet.py]: Maybe it should be GPU_TAG and CPU_TAG, now if we want to have CPU data we need to use LABEL_TAG.

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/mxnet.py]: For COCO detection pipeline we can't use this iterator anyway as we need to provide offsets as COCO iterator does.

[the remark this answers is not in the document]
Comment (awolant) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Right

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 7 above)
Comment (Kh4L) [reply dali/python/nvidia/dali/plugin/mxnet.py]: It is not about being GPU or CPU here.
MXNet's data input structure makes a distinction between Data and Label. Model implementations expect labels to be set in the `label` NDarray list and Input data in the `data` ndarray.
https://mxnet.incubator.apache.org/api/python/io/io.html#mxnet-io-helper-classes-functions

Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Understood.

File: docs/examples/pytorch/main.py (hunk 3 above)
Comment (klecki) [reply docs/examples/pytorch/main.py]: Done.

**Comment by klecki** [comment]
Build 40663

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 8]
```diff
@@ -99,6 +97,13 @@ def __init__(self,
         self._data_batches = [[None] for i in range(self._num_gpus)]
         self._counter = 0
         self._current_data_batch = 0
+        self._output_names_map = [x[0] for x in output_map]
+        self._output_categories_map = [x[1] for x in output_map]
```
Comment (Kh4L) [inline dali/python/nvidia/dali/plugin/mxnet.py]: You can do this here:
```
self._output_names_map, self._output_categories_map = zip(*output_map)
```

File: docs/examples/mxnet/mxnet-resnet50.ipynb [hunk 9]
```diff
@@ -200,22 +200,22 @@
...
+      "INFO:root:Epoch[0] Batch [300]\tSpeed: 4395.88 samples/sec\taccuracy=0.006074\n",
+      "INFO:root:Epoch[0] Batch [400]\tSpeed: 4384.70 samples/sec\taccuracy=0.011182\n",
+      "INFO:root:Epoch[0] Batch [500]\tSpeed: 4389.42 samples/sec\taccuracy=0.017441\n",
+      "INFO:root:Epoch[0] Batch [600]\tSpeed: 4382.10 samples/sec\taccuracy=0.026377\n",
+      "INFO:root:Epoch[0] Batch [700]\tSpeed: 4388.26 samples/sec\taccuracy=0.036611\n",
+      "INFO:root:Epoch[0] Batch [800]\tSpeed: 4383.51 samples/sec\taccuracy=0.047139\n",
+      "INFO:root:Epoch[0] Batch [900]\tSpeed: 4402.73 samples/sec\taccuracy=0.057686\n",
+      "INFO:root:Epoch[0] Batch [1000]\tSpeed: 4392.32 samples/sec\taccuracy=0.067861\n",
+      "INFO:root:Epoch[0] Batch [1100]\tSpeed: 4384.42 samples/sec\taccuracy=0.079248\n",
+      "INFO:root:Epoch[0] Batch [1200]\tSpeed: 4385.37 samples/sec\taccuracy=0.090088\n",
```
Comment (Kh4L) [inline docs/examples/mxnet/mxnet-resnet50.ipynb]: What is this speed performance drop? Different # of GPUs?

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 10]
```diff
@@ -99,6 +97,13 @@ def __init__(self,
...
+        self._output_categories = {DALIGenericIterator.DATA_TAG, DALIGenericIterator.LABEL_TAG}
+        assert set(self._output_categories_map) <= self._output_categories, \
```
Comment (Kh4L) [inline dali/python/nvidia/dali/plugin/mxnet.py]: Shouldn't it be also:
```suggestion
        assert len(set(self._output_categories_map)) <= len(self._output_categories), \
```

File: docs/examples/mxnet/mxnet-resnet50.ipynb (hunk 9 above)
Comment (JanuszL) [reply docs/examples/mxnet/mxnet-resnet50.ipynb]: Previous number may be generated somewhere else, but it would be good to double check if without that changes @klecki  gets the same numbers.

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 11]
```diff
@@ -130,37 +141,56 @@ def __next__(self):
...
             if self._data_batches[i][self._current_data_batch] is None:
-                d = [mx.nd.zeros(shape, mx.gpu(self._pipes[i].device_id), dtype = dtype) for shape, dtype in data_info]
-                l = [mx.nd.zeros(shape, mx.cpu(0), dtype = dtype) for shape, dtype in label_info]
+                mx_gpu_device = mx.gpu(self._pipes[i].device_id)
+                mx_cpu_device = mx.cpu(0)
+                from nvidia.dali.backend import TensorGPU
+                category_device = {key : [] for key in self._output_categories}
+                for category in self._output_categories:
+                    for t in category_tensors[category]:
+                        if type(t) is TensorGPU:
```
Comment (Kh4L) [inline dali/python/nvidia/dali/plugin/mxnet.py]: Why here?

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 12]
```diff
@@ -130,37 +141,56 @@ def __next__(self):
...
+            category_info[DALIGenericIterator.LABEL_TAG] = \
+                [(x.shape(), np.dtype(x.dtype())) for x in category_tensors[DALIGenericIterator.LABEL_TAG]]
+
             # If we did not yet allocate memory for that batch, do it now
             if self._data_batches[i][self._current_data_batch] is None:
-                d = [mx.nd.zeros(shape, mx.gpu(self._pipes[i].device_id), dtype = dtype) for shape, dtype in data_info]
-                l = [mx.nd.zeros(shape, mx.cpu(0), dtype = dtype) for shape, dtype in label_info]
+                mx_gpu_device = mx.gpu(self._pipes[i].device_id)
+                mx_cpu_device = mx.cpu(0)
+                from nvidia.dali.backend import TensorGPU
```
Comment (Kh4L) [inline dali/python/nvidia/dali/plugin/mxnet.py]: Why here?

File: dali/python/nvidia/dali/plugin/pytorch.py [hunk 13]
```diff
@@ -85,6 +88,8 @@ def __init__(self,
         self._data_batches = [[None, None] for i in range(self._num_gpus)]
         self._counter = 0
         self._current_data_batch = 0
+        self._output_categories = set(output_map)
+        assert len(set(output_map)) == len(output_map), "output_map names should be distinct"
```
Comment (Kh4L) [inline dali/python/nvidia/dali/plugin/pytorch.py]: Better swap L91 and L92?

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 8 above)
Comment (klecki) [reply dali/python/nvidia/dali/plugin/mxnet.py]: I never liked this zip(*something) tricks, as they can be harder to comprehend. I'll lave it as it is now.

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 10 above)
Comment (klecki) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Nope, because I want to change if there are only allowed elements in `_output_categories_map`, and not if the number of the categories is ok.

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 12 above)
Comment (klecki) [reply dali/python/nvidia/dali/plugin/mxnet.py]: It was the same in pytorch plugin, to limit the scope of the import.

File: dali/python/nvidia/dali/plugin/pytorch.py (hunk 13 above)
Comment (klecki) [reply dali/python/nvidia/dali/plugin/pytorch.py]: Done

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 10 above)
Comment (Kh4L) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Ok

File: dali/python/nvidia/dali/plugin/pytorch.py (hunk 2 above)
Comment (Kh4L) [reply dali/python/nvidia/dali/plugin/pytorch.py]: Sure, my comment targeted pytorch version only

**Comment by klecki** [comment]
Build 40666

File: docs/examples/mxnet/mxnet-resnet50.ipynb (hunk 9 above)
Comment (klecki) [reply docs/examples/mxnet/mxnet-resnet50.ipynb]: This is related to the machine I was using, I checked the before and after and it's consistent (even slower, cause almost everything was occupied, so I took something slower).
